src/ScheduleFunctions.cpp

```diff
@@ -79,6 +79,98 @@ bool contains_impure_call(const Expr &expr) {
     return is_not_pure.result;
 }
 
+// A mutator that performs a substitute operation only on either the values or the
+// arguments of Provide nodes.
+class SubstituteIn : public IRMutator {
+    const string &name;
+    const Expr &value;
+    bool calls;
+    bool provides;
+
+    using IRMutator::visit;
+
+    Stmt visit(const Provide *p) override {
+        if (!provides) {
+            return IRMutator::visit(p);
+        }
+        vector<Expr> args;
+        bool changed = false;
+        for (const Expr &i : p->args) {
+            args.push_back(substitute(name, value, i));
+            changed = changed || !args.back().same_as(i);
+        }
+        if (changed) {
+            return Provide::make(p->name, p->values, args, p->predicate);
+        } else {
+            return p;
+        }
+    }
+
+    Expr visit(const Call *op) override {
```

Comment:
If this is just mutating the LHS vs RHS, then this call visitor doesn't need to exist. Just call substitute in the Provide visitor.

Reply:
I think we really want to limit the substitutions to calls. We want only instances of coordinates inside calls to be clamped, not those outside.